Objects/lazyimportobject.c

```diff
@@ -0,0 +1,212 @@
+
+/* Lazy object implementation */
+
+#include "Python.h"
+#include "pycore_lazyimport.h"
+
+PyObject *
+PyLazyImport_NewModule(
+    PyObject *name, PyObject *globals, PyObject *locals, PyObject *fromlist, PyObject *level)
+{
+    PyLazyImportObject *m;
+    if (!name || !PyUnicode_Check(name) ||
+        !globals || !locals || !fromlist) {
+        PyErr_BadArgument();
+        return NULL;
+    }
+    if (level == NULL) {
+        level = PyLong_FromLong(0);
+        if (level == NULL) {
+            return NULL;
+        }
+    } else {
+        Py_INCREF(level);
+    }
+    m = PyObject_GC_New(PyLazyImportObject, &PyLazyImport_Type);
+    if (m == NULL) {
+        Py_DECREF(level);
+        return NULL;
+    }
+    m->lz_lazy_import = NULL;
+    Py_INCREF(name);
+    m->lz_name = name;
+    Py_INCREF(globals);
+    m->lz_globals = globals;
+    Py_INCREF(locals);
+    m->lz_locals = locals;
+    Py_INCREF(fromlist);
+    m->lz_fromlist = fromlist;
+    m->lz_level = level;
+    m->lz_resolved = NULL;
+    m->lz_resolving = NULL;
+    PyObject_GC_Track(m);
+    return (PyObject *)m;
+}
+
+PyObject *
+PyLazyImport_NewObject(PyObject *from, PyObject *name)
+{
+    PyLazyImportObject *m;
+    if (!from || !PyLazyImport_CheckExact(from) || !name || !PyUnicode_Check(name)) {
+        PyErr_BadArgument();
+        return NULL;
+    }
+    m = PyObject_GC_New(PyLazyImportObject, &PyLazyImport_Type);
+    if (m == NULL) {
+        return NULL;
+    }
+    PyLazyImportObject *lazy_from = (PyLazyImportObject *)from;
+    if (lazy_from->lz_fromlist != NULL && lazy_from->lz_fromlist != Py_None) {
+        PyObject *fromlist = PyTuple_New(1);
+        if (fromlist == NULL) {
+            return NULL;
+        }
+        Py_INCREF(name);
+        if (PyTuple_SetItem(fromlist, 0, name) < 0) {
+            return NULL;
+        }
+        PyObject *new_from = PyLazyImport_NewModule(
+            lazy_from->lz_name, lazy_from->lz_globals, lazy_from->lz_locals, fromlist, lazy_from->lz_level);
+        Py_DECREF(fromlist);
+        if (new_from == NULL) {
+            return NULL;
+        }
+        m->lz_lazy_import = new_from;
+    } else {
+        Py_INCREF(from);
+        m->lz_lazy_import = from;
+    }
```

Comment:
These snippet:

```py
from foo.bar import baz, thud
```

Translates to something like:

```
LOAD_CONST               0 (0)
LOAD_CONST               1 (('baz', 'thud'))
IMPORT_NAME              0 (foo.bar)
IMPORT_FROM              1 (baz)
STORE_NAME               1 (baz)
IMPORT_FROM              2 (thud)
STORE_NAME               2 (thud)
POP_TOP
```

We'd end up with two lazy objects that look like this:

```py
baz = LazyObject(
    from=LazyObject(
        name="foo.bar",
        fromlist=("baz", "thud"),
    ),
    name="baz",
)
thud = LazyObject(
    from=LazyObject(
        name="foo.bar",
        fromlist=("baz", "thud"),
    ),
    name="thud",
)
```

Because it's reusing the single `IMPORT_NAME` on each of the two `IMPORT_FROM`. At the time we resolve the imports, the `fromlist` (`("baz", "thud")`), would make it so accessing either `baz` or `thud` would trigger importing both submodules and not just one.

In this code, I'm making a copy of the inner lazy import with only a single element in its `fromlist` (the corresponding one of the current `IMPORT_FROM`); basically splitting the import as if it had originally been:

```py
from foo.bar import baz
from foo.bar import thud
```

At the end, we end up with:

```py
baz = LazyObject(
    from=LazyObject(
        name="foo.bar",
        fromlist=("baz",),
    ),
    name="baz",
)
thud = LazyObject(
    from=LazyObject(
        name="foo.bar",
        fromlist=("thud"),
    ),
    name="thud",
)
```

Which continues to work the exact same way, without having to load the unneeded submodules.